지역설정 화면 구현

Jeon0976 opened the pull request:
## Motivation ⍰

- 지역 화면 구현 

<br>

## Key Changes 🔑
- APP

  1. HomeSceneDIContainer 수정
       - Infra Service를 주입 받을 수 있는 `Dependencies` 구조체 생성
       - `DefaultRegionSettingUseCase` 생성가능한 메서드 생성
    
  2.`extetnsion`으로 Domain / Data / Presentation 영역 분리 

- Domain / Entities 
   1. MainRegion 엔티티 구현
       - `region` Enum,  `isSelected` Bool을 보유하고 있습니다. 
       - `isSelected`는 MainRegion 선택에 따라 SubRegion 값을 불러오기 위함입니다. 
   2. SubRegion 엔티티 구현
       - `region` String, `themeCount` Int?을 보유하고 있습니다.
       -  `themeCount` 값은 nil을 보장합니다. 이는 회원가입 로직간 count가 없는 화면이 있기 때문이며 해당 값을 불러오는 행위는 `RegionsSettingUseCase` 에서 진행합니다. 

- Domain / UseCase
   1. `RegionSettingUseCase` 구현
       - `loadMainRegions` Main 화면의 지역 값에 맞춰서 최초 화면을 불러오기 위해 파라미터 값을 받습니다. 현재 int형을 받지만, Region Enum 값 또는 String 값 중 효율적인 값으로 리팩토링 예정입니다. 
       - `loadSubRegions` region enum 파라미터를 받으며 해당 값을 통해 `fetchSubRegionsThemeCount` 에서 count를 호출합니다. 여기서도 count를 안 불러도 되는 화면이 있기때문에 호출할지 말지 분류할 수 있는 값과 로직을 추가할 예정입니다. 
       - `fetchSubRegionsThemeCount` 추후 repository 구현 시 연동 예정

- Presentation / Coordinator 
   - RegionSettingVC 구현 메서드 생성 

- Presentation / ViewModel 
   - Actions 구조체 추가
   - RxSwift / RxCocoa를 활용하여 DataBinding 
   
- Presentation / ViewController, View
   - Navigation Title 초기값 설정 및 외부에서 주입 받게 수정 필요

<br>

## To Reviewers 🙏🏻

- BaseViewController에서 DataBinding 함수를 사용하면 ViewDidLoad 데이터 스트림 구독 시점이 적절하지 않습니다. 
- 각각의 create static 함수 내부에서 DataBinding 하는게 어떨가 싶습니다. 

- Main에서 RegionSetting 화면으로 좌측에서 나오도록 커스텀 Push 기능 구현 예정입니다. 

<br>

## Linked Issue 🔗

-  #24

Jeon0976 committed:
[Feat]: Main / Sub Region Table 구현 (내부 코드 수정 예정)

Jeon0976 committed:
[Feat]: TableViewUpdate

Jeon0976 committed:
[Chore] Merge

Jeon0976 committed:
[Chore]: Merge

Jeon0976 committed:
Feat: RegionSetting UI 및 기능 구현

 - Domain: Entities / UseCase 구현

 - VC /ViewModel 구현

Jeon0976 commented on Nagaza/Sources/Domain/UseCases/RegionSettingUseCase.swift:
```diff
@@ -0,0 +1,27 @@
+        
+    }
+    
+    func loadMainRegions(with isSelectedIndex: Int = 0) -> [MainRegion] {
```
index가 아닌 SubRegion  "title" 값을 받아와서 해당 SubRegion title이 있는 MainRegion이 selected되도록 내부 코드를 수정하겠습니다.

Jeon0976 replied:
파라미터를 위에 subRegion title로 받는 `loadMainRegions` 함수를 하나 더 만들겠습니다!

Jeon0976 commented on Nagaza/Sources/Domain/UseCases/RegionSettingUseCase.swift:
```diff
@@ -0,0 +1,27 @@
+        return fetchSubRegionsThemeCount(with: mainRegion)
+    }
+    
+    private func fetchSubRegionsThemeCount(with mainRegion: Region) -> [SubRegion] {
```
main regions을 클릭 할 때마다 해당 main region 내부 subRegions count 개수를 불러들이는 api를 호출 할 예정인데, 
해당 useCase를 호출 한 ViewModel에서 이미 한 번 호출했던 main region을 호출하면 api 호출 안하고 내부 캐쉬 데이터로 return 할 수 있도록 수정할 예정입니다.  또한
 
해당 캐쉬는 useCase가 메모리에서 초기화 되면 삭제될 수 있도록 수정할 예정입니다.

Jeon0976 committed:
Feat: DIContainer & Coordinator 연결

Jeon0976 committed:
[Feat]: Home VC <-> RegionSettingVC 데이터 바인딩

 - Delegate 패턴을 사용하지 않고 의존성 주입만으로 child VC에서 present VC로 data 주입

Jeon0976 commented:
뒤로가기 아이콘, 애니메이션 추가, subRegion 클릭될때 animation 등 기타 UI 수정은 차차 진행하겠습니다.

Jeon0976 committed:
fix: 기타 중복으로 인해 서울 ~ 전라도 까지 selected 되는 문제 해결

 '기타'에서 '서울 기타', '경기도 기타'... 순으로 string값 수정

Minny27 commented on Nagaza/Sources/Application/DIContainer/AppDIContainer.swift:
```diff
@@ -15,7 +15,9 @@ final class AppDIContainer {
     
     func makeHomeSceneDIContainer() -> HomeSceneDIContainer {
-        return HomeSceneDIContainer()
+        let dependencise = HomeSceneDIContainer.Dependencies()
```
app 단에서 service를 주입하는 이유를 알 수 있을까요? 앱에서 직접적으로 data에 의존할 필요는 없을 거 같아서요
또한, VC에서 viewModel에 usecase를 주입하기 때문에 service를 주입할 필요가 없을 거 같습니다.

Minny27 committed:
Refactor: HomeUseCaseInterface를 준수하는 DefaultHomeUseCase로 변경

- HomeUseCase -> DefaultHomeUseCase로 이름 변경
- HomeUseCaseInterface라는 이름으로 프로토콜을 만들고 DefaultHomeUseCase가 준수

Minny27 committed:
Refactor: 홈 화면 추천 VC, 테마 컬렉션 VC를 UIView로 전환

- Base View에 layoutSubviews 오버라이딩 함수 추가
- 필요없는 파일 삭제

Minny27 commented on Nagaza/Sources/Presentation/Feature/TabBar/Home/RegionSetting/RegionSettingViewController.swift:
```diff
@@ -0,0 +1,172 @@
+//
+//  RegionSettingViewController.swift
```
Home 파일 뎁스가 View, RegionSetting으로 나뉘어서 View안에 넣는게 어떤가 싶네요

Minny27 commented on Nagaza/Sources/Presentation/Feature/Flow/HomeFlowCoordinator.swift:
```diff
@@ -43,6 +52,15 @@ final class HomeFlowCoordinator: Coordinator {
+        with subRegion: String,
+        didSelect: @escaping RegionSettingViewModelDidSelectAction
+    ) {
+        let vc = dependencies.makeRegionSettingViewController(with: subRegion, didSelect: didSelect)
```
subRange는 회원가입할 때 설정하거나 나중에 마이 페이지에서 변경할 수도 있을 거 같아서 상태 값을 UserDefault나 다른 방식으로 글로벌하게 저장하는게 어떤가 싶습니다.
서버로부터 따로 응답을 받아서 처리할 필요도 없기도 하고요

Jeon0976 replied:
해당 상태값은 UserDefault로 저장하고 UserDefaults 관리는 Data영역에서 할 수 있도록 설정예정입니당

Minny27 replied:
공유 자원이면 vc 간 데이터 전송 구현은 필요는 없을 거 같습니다.

Jeon0976 replied:
아 맞네요! data영역 해당 repository만 만들면 모든게 다 원큐로 해결가능하네용 
맞춰서 리펙토링 하겠습니다

Minny27 committed:
Refactor: UI 수정

- 오토 레이아웃 view.safeAreaLayoutGuide에 맞춤
- separator inset zero로 설정

Minny27 committed:
Refactor: mapButtonTapSubject bindViewModel 함수에 포함

- mapButtonItem을 View Controller에 선언
- map button 이벤트를 bind view model에 적용

Minny27 commented on Nagaza/Sources/Application/DIContainer/SceneDIContainer/HomeSceneDIContainer.swift:
```diff
@@ -0,0 +1,92 @@
+    
+    func makeRegionSettingViewController(
+        with subRegion: String,
+        didSelect: @escaping RegionSettingViewModelDidSelectAction
```
home에서 subject를 넘겨주고 makeRegionSettingViewController에서 구독하고
cell 버튼 클릭 발생했을 때 값을 emit하는 방식으로 하는 게 어떤가 싶긴하네요

Jeon0976 replied:
해당 방식도 괜찮을것같아요
기회가 될 때 subject로 리펙토링 해보겠습니다

Minny27 commented:
이전 develop merge 후 push하시고 merge하시면 될 거같습니다

고생하셨습니다. :)

Jeon0976 committed:
Merge branch 'develop' into Feature/LocationFilter
